### DISCUSSION TITLE
Need help defining early stopping behavior

### DISCUSSION THREAD

**Discussion by zachlim98** [explosion/spaCy#7450] [Help: Other Questions] [answered]
![image](https://user-images.githubusercontent.com/68678549/111256399-f83bac80-8653-11eb-90d9-8503ec1ad48f.png)

Love using Spacy 3.0 because it seems much cleaner but I need a little help interpreting the output. E would be epochs but what does # refer to? 

```
[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.4
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null
```

And given those, how does `patience` in config.cfg affect my early stopping behaviour? The default is set to 1600 so surely that isn't the number of epochs right? 

Thank you!

**Comment by polm** [comment, marked as answer]
Good question! The `#` column is the number of optimization steps (= batches processed), and `patience` is the number of steps to go without improvement before stopping.

**Reply by zachlim98** [reply]
Thanks for your reply! Just another question, if I were to set patience to 0 (i.e. turn off early stopping behavior), how would I go about setting the max number of iterations? In spacy 2.3.5., it was easy to set `n_iter` but I don't see a similar option in config? Unless I'm missing something. Thank you!

**Reply by polm** [reply]
`max_epochs` and `max_steps` both control the maximum number of iterations. With your config above, even if early stopping is never triggered it would stop after 20,000 steps.

**Reply by zachlim98** [reply]
Ah I see! So, in this case, if I need it to see my whole data at least once (assuming a 200,000 size train set), I would need to set the `max_steps` to at least the size of the train set?

**Reply by polm** [reply]
You could do that. Usually an epoch is defined as "seeing the all the data once", so the easiest thing would be to set the epochs.

**Reply by zachlim98** [reply]
Thanks so much for your help! Understand that a lot better now.

**Comment by yazeedp** [comment]
Can someone help out on the below query:
While retraining the NER using the CLI, are there any parameters in the cfg file we can include to change the frequency or timing on when the weights are updated to the model file on disk.And Can You also help on what the default setting is relating to when updated weights are saved to the model path.

**Reply by polm** [reply]
@yazeedp Please open a new Discussion instead of asking on this one.
